test/metabase/test_runner/assert_exprs/approximately_equal_test.clj

```diff
@@ -0,0 +1,124 @@
+               (when-not (= expected actual-str)
+                 (list 'not= expected (symbol "#t") actual-str))))}
+          "2022-07-14"
+          #t "2022-07-14"))
+
+  (is (=? {[String String]
+           (fn [_next-method ^String expected ^String actual]
+             (when-not (zero? (.compareToIgnoreCase expected actual))
+               (list 'not (list 'zero? (list '.compareToIgnoreCase expected actual)))))}
+          {:a "AbC"}
+          {:a "abc", :b 100})))
+
+(deftest exactly-test
+  (testing "#exactly"
+    (is (=? {:a 1}
+            {:a 1, :b 2}))
+    (testing "Fail when things are not exactly the same, as if by `=`"
+      ;; convert to a string because otherwise two regexes aren't equal and I didn't want to use `?=` to test itself.
+      (is (= `(~'not (~'= ~(symbol "#exactly") {:a 1} {:a 1, :b 2}))
+             (approximately-equal/=?-diff #exactly {:a 1} {:a 1, :b 2})))
+      (testing "Inside a map"
+        (is (= `{:b (~'not (~'= ~(symbol "#exactly") {:a 1} {:a 1, :b 2}))}
+               (approximately-equal/=?-diff {:a 1, :b #exactly {:a 1}}
+                                            {:a 1, :b {:a 1, :b 2}})))))
+    (testing "Should pass when things are exactly the same as if by `=`"
+      (is (nil? (approximately-equal/=?-diff #exactly 2 2)))
+      (is (=? #exactly 2
+              2))
+      (testing "should evaluate args"
+        (is (=? #exactly (+ 1 1)
+                2)))
+      (testing "Inside a map"
+        (is (=? {:a 1, :b #exactly 2}
+                {:a 1, :b 2}))))))
+
+(deftest schema-test
+  (testing "#schema"
+    (is (=? #schema {:a s/Int}
+            {:a 1}))
+    (testing "Nested inside a collection"
+      (is (=? {:a 1, :b #schema {s/Keyword s/Int}}
+              {:a 1, :b {}}))
+      (is (=? {:a 1, :b #schema {s/Keyword s/Int}}
+              {:a 1, :b {:c 2}}))
+      (is (=? {:a 1, :b #schema {s/Keyword s/Int}}
+              {:a 1, :b {:c 2, :d 3}})))
+    (testing "failures"
+      ;; serialize these to strings and read them back out because Schema actually returns weird classes like
+      ;; ValidationError or whatever that aren't equal to their printed output
+      (is (= '{:a (not (integer? 1.0))}
+             (read-string (pr-str (approximately-equal/=?-diff #schema {:a s/Int} {:a 1.0})))))
+      (testing "Inside a collection"
+        (is (= '{:b {:c (not (integer? 2.0))}}
+               (read-string (pr-str (approximately-equal/=?-diff {:a 1, :b #schema {:c s/Int}}
+                                                                 {:a 1, :b {:c 2.0}})))))))))
+
+(deftest approx-test
+  (testing "#approx"
+    (is (=? #approx [1.5 0.1]
+            1.51))
+    (testing "Nested inside a collection"
+      (is (=? {:a 1, :b #approx [1.5 0.1]}
+              {:a 1, :b 1.51})))
+    ;; failures below render stuff to strings so we can see it the way it will look in test failures with it's nice
+    ;; comment and what not
```

Comment:
```suggestion
    ;; failures below render stuff to strings so we can see it the way it will look in test failures with its nice
    ;; comment and whatnot
```

:shrug: